File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        vector<int> sum_;
        int sum=0;
        int cnt=0;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
            if(sum==k){
                cnt++;
            }
            cout<<"sum:"<<sum<<endl;
            sum_.push_back(sum);
            

        }
        
        for(int i = 0; i < nums.size(); i++) {
            for(int j = i + 1; j < nums.size(); j++) {

                if(sum_[j] - sum_[i] == k) {
                    cnt++;
                }
            }
        }

        return cnt;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        # prefix = prefix + nums[i]

        # diff = prefix - 3

        # if diff exists in map:
        #     cnt += map[diff]

        # map[prefix] += 1
        prefix=[nums[0]]
        for i in range(1,len(nums)):
            prefix.append(prefix[-1]+nums[i])
        hm={0:1}
        cnt=0
        # print(prefix)
        for i in range(len(nums)):
            diff=prefix[i]-k
            if diff in hm:
                cnt+=hm[diff]
            hm[prefix[i]]=hm.get(prefix[i], 0) + 1

        return cnt
